pennylane/beta/tapes/cv_param_shift.py

```diff
@@ -0,0 +1,458 @@
+
+    This tape supports analytic gradients of photonic circuits that satisfy
+    the following constraints with regards to measurements:
+
+    * Expectation values are restricted to observables that are first- and
+      second-order in :math:`\hat{x}` :math:`\hat{p}` only.
+      This includes :class:`~.X`, :class:`~.P`, :class:`~.QuadOperator`,
+      :class:`~.PolyXP`, and :class:`~.NumberOperator`.
+
+      For second-order observables, the device **must support** :class:`~.PolyXP`.
+
+    * Variances are restricted to observables that are first-order
+      in :math:`\hat{x}` :math:`\hat{p}` only. This includes :class:`~.X`, :class:`~.P`,
+      :class:`~.QuadOperator`, and *some* parameter values of :class:`~.PolyXP`.
+
+      The device **must support** :class:`~.PolyXP`.
+
+    Fock state probabilities (tapes that return :func:`~pennylane.probs` or
+    expectation values of :class:`~.FockStateProjector`) are not supported.
+
+    In addition, the tape operations must fulfill the following requirements:
+
+    * Only Gaussian operations are differentiable.
+
+    * Non-differentiable Fock states and Fock operations may *precede* all differentiable Gaussian,
+      operations. For example, the following is permissible:
+
+      .. code-block:: python
+
+          with CVParamShiftTape() as tape:
+              # Non-differentiable Fock operations
+              qml.FockState(2, wires=0)
+              qml.Kerr(0.654, wires=1)
+
+              # differentiable Gaussian operations
+              qml.Displacement(0.6, wires=0)
+              qml.Beamsplitter(0.5, 0.1, wires=[0, 1])
+              expval(qml.NumberOperator(0))
+
+          tape.trainable_params = {2, 3, 4}
+
+    * If a Fock operation succeeds a Gaussian operation, the Fock operation must
+      not contribute to any measurements. For example, the following is allowed:
+
+      .. code-block:: python
+
+          with CVParamShiftTape() as tape:
+              qml.Displacement(0.6, wires=0)
+              qml.Beamsplitter(0.5, 0.1, wires=[0, 1])
+              qml.Kerr(0.654, wires=1)  # there is no measurement on wire 1
+              expval(qml.NumberOperator(0))
+
+          tape.trainable_params = {0, 1, 2}
+
+    If any of the above constraints are not followed, the tape cannot be differentiated
+    via the CV parameter-shift rule. Please use numerical differentiation instead:
+
+    >>> tape.jacobian(dev, method="numeric")
+    """
+
+    def _grad_method(self, idx, use_graph=True, default_method="F"):
+        op = self._par_info[idx]["op"]
+
+        if op.grad_method in (None, "F"):
+            return op.grad_method
+
+        if op.grad_method != "A":
+            raise ValueError(f"Operation {op} has unknown gradient method {op.grad_method}")
+
+        if not use_graph:
+            raise ValueError(
+                "The CV parameter-shift rule must always use the "
+                "graph to determine operation gradient methods"
+            )

```

Comment:
We _could_ remove `use_graph` from the signature, but this would violate the [Liskov substitution principle](https://en.wikipedia.org/wiki/Liskov_substitution_principle), so thought it was best to leave it as is (and just catch the user _trying_ to not use the graph!

There might also be a nice way of avoiding the graph? I just couldn't think of one here 😆